### PR TITLE
Incorrect output shape of Gather operation for some models

### PR DESCRIPTION
**Pull request by vgavrilo**
Root cause analysis: 
When `params` and `indices` inputs of the operation `Gather` have output ranks less than 4, and an output rank of `Gather` is greater or equal to 4, and a source model is a TF model, then the generated IR contains `Transpose` layer after such `Gather`. But this inserted `Transpose` is not needed in such case, because then TF `Gather` and OV `Gather` gives identical shapes and values.

Solution:
Add `node['reinterp_shape'] = True` in the infer function of the MO operation `Gather` for the above mentioned case.

Ticket: CVS 65857

Code:
* [x]  Comments
* [x]  Code style
* [x]  Transformation generates reshape-able IR - N/A
* [x]  Transformation preserves original framework node names - N/A

Validation:
* [x]  Unit tests
* [x]  Framework operation tests  - N/A
* [x]  Transformation tests - N/A
* [x]  Model Optimizer IR Reader check - N/A

Documentation:
* [x]  Supported frameworks operations list - N/A
* [x]  Guide on how to convert the **public** model - N/A
* [x]  User guide update - N/A

### CODE REVIEW

File: model-optimizer/extensions/ops/gather.py [hunk 1]
```diff
@@ -98,6 +98,9 @@ def infer(node: Node):
         else:
             node.out_port(0).data.set_shape(out_shape)
 
+        if max(len(data_shape), len(indices_shape)) < 4 and len(out_shape) >= 4:
+            node['reinterp_shape'] = True
```
Comment (popovaan) [inline model-optimizer/extensions/ops/gather.py:102]: Can you please add a comment why we need this attribute.

File: model-optimizer/extensions/ops/gather.py [hunk 2]
```diff
@@ -98,6 +98,9 @@ def infer(node: Node):
         else:
             node.out_port(0).data.set_shape(out_shape)
 
+        if max(len(data_shape), len(indices_shape)) < 4 and len(out_shape) >= 4:
```
Comment (popovaan) [inline model-optimizer/extensions/ops/gather.py:101]: Please add uint test of partial inference for this case in unit_tests/extensions/ops/gather_test.py

**Review by popovaan — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/ops/gather.py (hunk 1 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/gather.py:110]: Done.

File: model-optimizer/extensions/ops/gather.py (hunk 2 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/gather.py:109]: Done.

**Comment by vgavrilo** [comment]
@pavel-esir, @popovaan, @sadolini, @rkazants, @jane-intel, please, see this PR (https://github.com/openvinotoolkit/openvino/pull/8899).

**Review by popovaan — APPROVED** [review]

File: model-optimizer/extensions/ops/gather.py [hunk 3]
```diff
@@ -98,6 +98,17 @@ def infer(node: Node):
...
 
+        # When 'params' and 'indices' inputs of the operation Gather have output ranks less than 4,
+        # and an output rank of 'Gather' is greater or equal to 4, and a source model is a TF model,
+        # then the generated IR contains Transpose layer after such Gather.
+        #
+        # But this inserted Transpose is not needed in such case, because then TF 'Gather' and OV 'Gather'
+        # gives identical shapes and values.
+        #
+        # To fix it, we need set attribute 'reinterp_shape' of such 'Gather' as True.
+        if max(len(data_shape), len(indices_shape)) < 4 and len(out_shape) >= 4:
```
Comment (jane-intel) [inline model-optimizer/extensions/ops/gather.py:109]: it should always be marked -- not only in 4 and 5D case

Comment (jane-intel) [reply model-optimizer/extensions/ops/gather.py:109]: Attributed gather should be marked the same

**Review by jane-intel — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/ops/gather.py (hunk 3 above)
Comment (vgavrilo) [reply model-optimizer/extensions/ops/gather.py]: Fixed.

**Review by jane-intel — APPROVED** [review]

**Pull request merged**